File: dimos/web/relay_bridge/manifest.py
# Copyright 2026 Dimensional Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Manifest domain model: the Python mirror of web/shared/manifest.ts.

Pinned by the golden vectors in web/shared/fixtures/manifests.json (tested
from both pytest and deno test). The transport (protocol.py) carries the
manifest as one opaque dict; this module is the single owner of its
structure and domain rules: version gate, bounded unique ids, positive
rates, panel/layout/pages references that resolve, and kind-specific panel
rules (video, map2d, teleop, chat, stats).

Manifest v1 is frozen. Additive changes (new panel kinds, new params) ride
the existing shape: unknown keys and kinds pass through validation.
Breaking changes bump `version`, and the Cockpit refuses politely
(unsupported_version).
"""

import json
import re
import sys
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, TypeAdapter, ValidationError

Delivery = Literal["latest", "reliable"]
# Flow direction seen from the viewer: rx = robot->viewer; tx (teleop, chat)
# arrives with later tickets.
Dir = Literal["rx", "tx"]
# Generic-publish policy for a tx channel: "none" (default; specialized
# protocol paths like teleop only), "shared" (any authorized viewer may
# publish), "exclusive" (requires the per-robot publisher lease, W8).
# Transport delivery and publish authorization are independent: reliable
# does not imply publishable.
Publish = Literal["none", "shared", "exclusive"]

# The only manifest version this build understands.
MANIFEST_VERSION = 1

# Bound for channel/panel ids, encodings, and panel kinds.
MAX_MANIFEST_ID_LEN = 64

# Channel ids with this prefix belong to protocol control (e.g. @control)
# and can never be declared by a manifest.
RESERVED_CHANNEL_PREFIX = "@"


class ManifestError(ValueError):
    """`code` is the machine-readable reason, pinned by the golden vectors."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(f"{code}: {message}")


class _ManifestModel(BaseModel):
    # strict + allow_inf_nan=False, mirroring _WireModel in protocol.py
    # (which imports ChannelSpec from here, so this base cannot be shared).
    model_config = ConfigDict(strict=True, allow_inf_nan=False)


class ChannelSpec(_ManifestModel):
    """One robot<->viewer stream (see ChannelSpec in manifest.ts).

    Field names and order are the wire names/order, hence the camelCase.
    `params` carries encoder settings (e.g. jpeg quality); absent fields
    normalize to their defaults. publish/requiredScope gate generic
    publishing (W7); requiredScope accepts explicit null so a normalized
    manifest parses idempotently, while a null publish is a shape error
    (like dir).
    """

    ch: str
    dir: Dir = "rx"
    encoding: str
    delivery: Delivery
    maxHz: int | float
    params: dict[str, Any] = Field(default_factory=dict)
    publish: Publish = "none"
    requiredScope: str | None = None


class PanelSpec(_ManifestModel):
    """kind is the panel-component registry key; channels lists the channel
    ids the panel consumes in kind-defined slot order (e.g. map2d: costmap,
    then the optional pose overlay). title "" means untitled (the frame
    falls back to the panel id)."""

    id: str
    kind: str
    title: str = ""
    channels: list[str]
    params: dict[str, Any] = Field(default_factory=dict)


class Manifest(_ManifestModel):
    version: int | float
    channels: list[ChannelSpec]
    panels: list[PanelSpec] = Field(default_factory=list)
    # Recursive layout tree: a leaf is a panel-id string; row/col nodes are
    # {"row"/"col": [children], "shares"?: [floats]}. Modeled as Any and
    # validated by the hand-rolled walker below so structural failures map
    # to invalid_layout AFTER the panel/kind rules (document order, like
    # manifest.ts) and normalization can drop unknown node keys while
    # preserving `shares` presence. None = no layout authored.
    layout: Any = None
    # Panel ids rendered as full-page tabs. Kept out of the model input in
    # parse_manifest so its shape errors fire in the pages phase (after
    # layout), mirroring manifest.ts.
    pages: list[str] = Field(default_factory=list)


_MANIFEST_TA: TypeAdapter[Manifest] = TypeAdapter(Manifest)


def _bounded_id(s: str) -> bool:
    return 1 <= len(s) <= MAX_MANIFEST_ID_LEN


# "Supported JSON encoding" for generic publish is the family-name rule
# (json.v1, text.json.v1, ...): the manifest layer cannot see the codec
# registries, so real decodability is enforced at authoring time.
_JSON_ENCODING_RE = re.compile(r"(^|\.)json\.v[0-9]+$")


def _validate_layout_node(node: Any, panel_ids: set[str], seen: set[str]) -> Any:
    """Depth-first layout validation + rebuild (see validateLayout in
    manifest.ts). A node's own structure (row/col exclusivity, children,
    shares) is checked before its children; string leaves check unknown
    before duplicate. `seen` is shared with the pages loop so one panel can
    be placed only once across the whole cockpit. Unknown node keys are
    dropped by the rebuild (global unknown-key policy)."""
    if isinstance(node, str):
        if node not in panel_ids:
            raise ManifestError("unknown_layout_panel", f"layout names unknown panel {node}")
        if node in seen:
            raise ManifestError("duplicate_layout_panel", f"panel {node} placed more than once")
        seen.add(node)
        return node
    if not isinstance(node, dict):
        raise ManifestError("invalid_layout", "layout node must be a panel id or a row/col object")
    is_row = "row" in node
    if is_row == ("col" in node):
        raise ManifestError("invalid_layout", "layout node needs exactly one of row/col")
    children = node["row"] if is_row else node["col"]
    if not isinstance(children, list) or not children:
        raise ManifestError("invalid_layout", "row/col children must be a non-empty array")
    shares = None
    if "shares" in node:
        shares = node["shares"]
        # bool is excluded explicitly (Python bool is an int); the float64
        # bound mirrors JS Number.isFinite, which rejects huge JSON ints
        # that parse to Infinity there.
        if (
            not isinstance(shares, list)
            or len(shares) != len(children)
            or any(
                isinstance(s, bool)
                or not isinstance(s, (int, float))
                or not 0 < s <= sys.float_info.max
                for s in shares
            )
        ):
            raise ManifestError("invalid_layout", "shares must be positive numbers, one per child")
    rebuilt = [_validate_layout_node(child, panel_ids, seen) for child in children]
    out: dict[str, Any] = {"row" if is_row else "col": rebuilt}
    if shares is not None:
        out["shares"] = shares
    return out


def parse_manifest(data: Any) -> Manifest:
    """Validated manifest from parsed JSON (or any untrusted value); raises
    ManifestError. Absent dir/params/title/layout/pages normalize to
    "rx"/{}/""/None/[]; unknown keys are dropped (forward compatibility
    with newer bridges). Mirrors parseManifest() in manifest.ts: shape,
    then domain rules in document order, so both sides report the same code
    (pinned by fixtures). The version gate runs before the channel/panel
    shape checks so a structurally-alien future manifest still reports
    unsupported_version, which the Cockpit turns into a polite notice.
    """
    # Through JSON, so the hand checks and the strict model see the same
    # JSON-shaped values a peer would put on the wire (strict python-mode
    # validation wants model instances for nested fields, but callers hold
    # parsed-JSON dicts).
    try:
        raw = json.loads(json.dumps(data))
    except (TypeError, ValueError) as e:
        raise ManifestError("invalid_shape", str(e)) from e
    if not isinstance(raw, dict):
        raise ManifestError("invalid_shape", "manifest must be an object")
    version = raw.get("version")
    if isinstance(version, bool) or not isinstance(version, (int, float)):
        raise ManifestError("invalid_shape", "manifest version must be a number")
    if version != MANIFEST_VERSION:
        raise ManifestError("unsupported_version", f"manifest version {version} is not supported")
    try:
        manifest = _MANIFEST_TA.validate_json(
            json.dumps({k: v for k, v in raw.items() if k != "pages"})
        )
    except ValidationError as e:
        raise ManifestError("invalid_shape", str(e)) from e

    ch_ids: dict[str, ChannelSpec] = {}
    for spec in manifest.channels:
        if not _bounded_id(spec.ch):
            raise ManifestError(
                "invalid_channel_id", f"channel id must be 1..{MAX_MANIFEST_ID_LEN} chars"
            )
        if spec.ch.startswith(RESERVED_CHANNEL_PREFIX):
            raise ManifestError(
                "reserved_channel_id",
                f"channel ids beginning with {RESERVED_CHANNEL_PREFIX} are reserved "
                "for protocol control",
            )
        if spec.ch in ch_ids:
            raise ManifestError("duplicate_channel_id", f"duplicate channel {spec.ch}")
        ch_ids[spec.ch] = spec
        if not _bounded_id(spec.encoding):
            raise ManifestError(
                "invalid_encoding", f"encoding must be 1..{MAX_MANIFEST_ID_LEN} chars"
            )
        # Bounded to float64: a JSON integer beyond that range is exact in
        # Python but Infinity through JS JSON.parse, and manifest.ts rejects
        # it here with this code. Non-finite floats are already shape-rejected
        # (and isfinite() would raise OverflowError on huge ints).
        if not 0 < spec.maxHz <= sys.float_info.max:
            raise ManifestError("invalid_max_hz", f"maxHz for {spec.ch} must be a positive number")
        # Generic-publish rules. The manifest layer accepts "exclusive" (only
        # authoring and the bridge reject it until the lease ticket, W8).
        if spec.dir == "rx" and spec.publish != "none":
            raise ManifestError("invalid_publish", f"rx channel {spec.ch} cannot declare publish")
        if spec.publish != "none" and spec.delivery != "reliable":
            raise ManifestError("invalid_publish", f"publish channel {spec.ch} must be reliable")
        if spec.publish != "none" and not _JSON_ENCODING_RE.search(spec.encoding):
            raise ManifestError(
                "invalid_publish", f"publish channel {spec.ch} needs a JSON encoding"
            )
        if spec.requiredScope is not None and spec.publish == "none":
            raise ManifestError(
                "invalid_scope", f"channel {spec.ch} scope requires a publish policy"
            )
        if spec.requiredScope is not None and not _bounded_id(spec.requiredScope):
            raise ManifestError(
                "invalid_scope", f"scope for {spec.ch} must be 1..{MAX_MANIFEST_ID_LEN} chars"
            )

    panel_ids: set[str] = set()
    for panel in manifest.panels:
        if not (_bounded_id(panel.id) and _bounded_id(panel.kind)):
            raise ManifestError(
                "invalid_panel", f"panel id/kind must be 1..{MAX_MANIFEST_ID_LEN} chars"
            )
        if panel.id in panel_ids:
            raise ManifestError("duplicate_panel_id", f"duplicate panel {panel.id}")
        panel_ids.add(panel.id)
        for ch in panel.channels:
            if ch not in ch_ids:
                raise ManifestError(
                    "unknown_panel_channel", f"panel {panel.id} wants undeclared channel {ch}"
                )
        # Kind-specific rules (incl. dir=rx for viewer-rendered channels);
        # unknown kinds stay unvalidated (forward compatibility with newer
        # bridges).
        if panel.kind == "video":
            if len(panel.channels) != 1:
                raise ManifestError(
                    "invalid_video_panel", f"video panel {panel.id} must bind exactly one channel"
                )
            bound = ch_ids[panel.channels[0]]
            if bound.encoding != "jpeg.v1" or bound.delivery != "latest" or bound.dir != "rx":
                raise ManifestError(
                    "invalid_video_panel",
                    f"video panel {panel.id} needs a jpeg.v1 latest rx channel",
                )
        if panel.kind == "map2d":
            # channels[0] is the costmap; channels[1] (optional) the pose overlay.
            if len(panel.channels) not in (1, 2):
                raise ManifestError(
                    "invalid_map2d_panel",
                    f"map2d panel {panel.id} must bind one or two channels",
                )
            costmap = ch_ids[panel.channels[0]]
            if (
                costmap.encoding != "costmap.zlib.v1"
                or costmap.delivery != "latest"
                or costmap.dir != "rx"
            ):
                raise ManifestError(
                    "invalid_map2d_panel",
                    f"map2d panel {panel.id} needs a costmap.zlib.v1 latest rx channel first",
                )
            if len(panel.channels) == 2:
                pose = ch_ids[panel.channels[1]]
                if pose.encoding != "pose.json.v1" or pose.dir != "rx":
                    raise ManifestError(
                        "invalid_map2d_panel",
                        f"map2d panel {panel.id} pose channel must be a pose.json.v1 rx channel",
                    )
        if panel.kind == "teleop":
            if len(panel.channels) != 1:
                raise ManifestError(
                    "invalid_teleop_panel",
                    f"teleop panel {panel.id} must bind exactly one channel",
                )
            cmd = ch_ids[panel.channels[0]]
            if cmd.encoding != "twist.json.v1" or cmd.delivery != "latest" or cmd.dir != "tx":
                raise ManifestError(
                    "invalid_teleop_panel",
                    f"teleop panel {panel.id} needs a twist.json.v1 latest tx channel",
                )
        if panel.kind == "chat":
            # channels: the text input (publish tx), the messages, the idle
            # flag, the push-to-talk audio (publish tx).
            if len(panel.channels) != 4:
                raise ManifestError(
                    "invalid_chat_panel", f"chat panel {panel.id} must bind four channels"
                )
            text, messages, idle, audio = (ch_ids[ch] for ch in panel.channels)
            if (
                text.encoding != "text.json.v1"
                or text.delivery != "reliable"
                or text.dir != "tx"
                or text.publish != "shared"
            ):
                raise ManifestError(
                    "invalid_chat_panel",
                    f"chat panel {panel.id} needs a text.json.v1 reliable shared tx channel first",
                )
            if (
                messages.encoding != "chat.json.v1"
                or messages.delivery != "reliable"
                or messages.dir != "rx"
            ):
                raise ManifestError(
                    "invalid_chat_panel",
                    f"chat panel {panel.id} needs a chat.json.v1 reliable rx channel second",
                )
            if idle.encoding != "json.v1" or idle.delivery != "latest" or idle.dir != "rx":
                raise ManifestError(
                    "invalid_chat_panel",
                    f"chat panel {panel.id} needs a json.v1 latest rx channel third",
                )
            if (
                audio.encoding != "audio.json.v1"
                or audio.delivery != "reliable"
                or audio.dir != "tx"
                or audio.publish != "shared"
            ):
                raise ManifestError(
                    "invalid_chat_panel",
                    f"chat panel {panel.id} needs an audio.json.v1 reliable shared tx "
                    "channel fourth",
                )
        if panel.kind == "stats":
            if len(panel.channels) != 1:
                raise ManifestError(
                    "invalid_stats_panel", f"stats panel {panel.id} must bind exactly one channel"
                )
            stats = ch_ids[panel.channels[0]]
            if stats.encoding != "stats.json.v1" or stats.delivery != "latest" or stats.dir != "rx":
                raise ManifestError(
                    "invalid_stats_panel",
                    f"stats panel {panel.id} needs a stats.json.v1 latest rx channel",
                )

    seen: set[str] = set()
    if manifest.layout is not None:
        manifest.layout = _validate_layout_node(manifest.layout, panel_ids, seen)

    raw_pages = raw.get("pages", [])
    if not isinstance(raw_pages, list) or any(not isinstance(p, str) for p in raw_pages):
        raise ManifestError("invalid_shape", "bad pages shape")
    for page_id in raw_pages:
        if page_id not in panel_ids:
            raise ManifestError("unknown_page_panel", f"pages names unknown panel {page_id}")
        if page_id in seen:
            raise ManifestError("duplicate_layout_panel", f"panel {page_id} placed more than once")
        seen.add(page_id)
    manifest.pages = raw_pages

    return manifest

File: web/shared/manifest.ts
// Manifest domain model shared by the relay and the Cockpit. Mirrored in
// Python at dimos/web/relay_bridge/manifest.py and pinned by the golden
// vectors in ./fixtures/manifests.json (tested from both deno test and
// pytest).
//
// The transport (protocol.ts) carries the manifest as one opaque record;
// this module is the single owner of its structure and domain rules:
// version gate, bounded unique ids, positive rates, panel/layout/pages
// references that resolve, and kind-specific panel rules (video, map2d).
//
// Manifest v1 is frozen. Additive changes (new panel kinds, new params)
// ride the existing shape: unknown keys and kinds pass through validation.
// Breaking changes bump `version`, and the Cockpit refuses politely
// (unsupported_version).

export type Delivery = "latest" | "reliable";
export type Dir = "rx" | "tx";
// Generic-publish policy for a tx channel: "none" (default; specialized
// protocol paths like teleop only), "shared" (any authorized viewer may
// publish), "exclusive" (requires the per-robot publisher lease, W8).
// Transport delivery and publish authorization are independent: reliable
// does not imply publishable.
export type Publish = "none" | "shared" | "exclusive";

// One robot<->viewer stream: dir is the flow direction seen from the viewer
// (rx = robot->viewer; tx arrives with teleop/chat), encoding names the
// payload format (e.g. jpeg.v1, pose.json.v1), delivery picks the relay's
// forwarding policy, maxHz is the bridge's advertised send cap (for a
// publish channel: the aggregate accepted publish rate), params are encoder
// settings (informational for viewers). publish/requiredScope gate generic
// publishing; requiredScope names the operator scope a remote relay demands
// (null = none; the local relay never checks scopes).
export interface ChannelSpec {
  ch: string;
  dir: Dir;
  encoding: string;
  delivery: Delivery;
  maxHz: number;
  params: Record<string, unknown>;
  publish: Publish;
  requiredScope: string | null;
}

// kind is the panel-component registry key; channels lists the channel ids
// the panel consumes in kind-defined slot order (e.g. map2d: costmap, then
// the optional pose overlay). title "" means untitled (the frame falls back
// to the panel id).
export interface PanelSpec {
  id: string;
  kind: string;
  title: string;
  channels: string[];
  params: Record<string, unknown>;
}

// Recursive layout tree: a leaf is a panel id; row/col split the space with
// flex shares (absent shares = equal split, a renderer default that is
// deliberately not baked into normalized data).
export interface RowNode {
  row: LayoutNode[];
  shares?: number[];
}
export interface ColNode {
  col: LayoutNode[];
  shares?: number[];
}
export type LayoutNode = string | RowNode | ColNode;

export interface Manifest {
  version: 1;
  channels: ChannelSpec[];
  panels: PanelSpec[];
  /** null = no layout authored; the Cockpit falls back to a simple row. */
  layout: LayoutNode | null;
  /** Panel ids rendered as full-page tabs instead of grid cells. */
  pages: string[];
}

/** The only manifest version this build understands. */
export const MANIFEST_VERSION = 1;

/** Bound for channel/panel ids, encodings, and panel kinds. */
export const MAX_MANIFEST_ID_LEN = 64;

/** Channel ids with this prefix belong to protocol control (e.g. @control)
 * and can never be declared by a manifest. */
export const RESERVED_CHANNEL_PREFIX = "@";

export class ManifestError extends Error {
  constructor(readonly code: string, message: string) {
    super(`${code}: ${message}`);
    this.name = "ManifestError";
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

// Raw (pre-normalization) spec shapes: dir/params/title may be absent.
// requiredScope also accepts explicit null so a normalized manifest parses
// idempotently (parse(parse(m)) == parse(m)); publish does not (null is a
// shape error, like dir).
interface RawChannelSpec {
  ch: string;
  dir?: Dir;
  encoding: string;
  delivery: Delivery;
  maxHz: number;
  params?: Record<string, unknown>;
  publish?: Publish;
  requiredScope?: string | null;
}

interface RawPanelSpec {
  id: string;
  kind: string;
  title?: string;
  channels: string[];
  params?: Record<string, unknown>;
}

function isChannelSpec(value: unknown): value is RawChannelSpec {
  return (
    isRecord(value) &&
    typeof value.ch === "string" &&
    (value.dir === undefined || value.dir === "rx" || value.dir === "tx") &&
    typeof value.encoding === "string" &&
    (value.delivery === "latest" || value.delivery === "reliable") &&
    typeof value.maxHz === "number" &&
    (value.params === undefined || isRecord(value.params)) &&
    (value.publish === undefined || value.publish === "none" || value.publish === "shared" ||
      value.publish === "exclusive") &&
    (value.requiredScope === undefined || value.requiredScope === null ||
      typeof value.requiredScope === "string")
  );
}

function isPanelSpec(value: unknown): value is RawPanelSpec {
  return (
    isRecord(value) &&
    typeof value.id === "string" &&
    typeof value.kind === "string" &&
    (value.title === undefined || typeof value.title === "string") &&
    Array.isArray(value.channels) &&
    value.channels.every((ch) => typeof ch === "string") &&
    (value.params === undefined || isRecord(value.params))
  );
}

function boundedId(s: string): boolean {
  return s.length >= 1 && s.length <= MAX_MANIFEST_ID_LEN;
}

function dirOf(spec: RawChannelSpec): Dir {
  return spec.dir ?? "rx";
}

function publishOf(spec: RawChannelSpec): Publish {
  return spec.publish ?? "none";
}

function scopeOf(spec: RawChannelSpec): string | null {
  return spec.requiredScope ?? null;
}

// "Supported JSON encoding" for generic publish is the family-name rule
// (json.v1, text.json.v1, ...): the manifest layer cannot see the codec
// registries, so real decodability is enforced at authoring time.
const JSON_ENCODING_RE = /(^|\.)json\.v[0-9]+$/;

/**
 * Depth-first layout validation + rebuild. A node's own structure (row/col
 * exclusivity, children, shares) is checked before its children; string
 * leaves check unknown before duplicate. `seen` is shared with the pages
 * loop so one panel can be placed only once across the whole cockpit.
 * Unknown node keys are dropped by the rebuild (global unknown-key policy).
 */
function validateLayout(node: unknown, panelIds: Set<string>, seen: Set<string>): LayoutNode {
  if (typeof node === "string") {
    if (!panelIds.has(node)) {
      throw new ManifestError("unknown_layout_panel", `layout names unknown panel ${node}`);
    }
    if (seen.has(node)) {
      throw new ManifestError("duplicate_layout_panel", `panel ${node} placed more than once`);
    }
    seen.add(node);
    return node;
  }
  if (!isRecord(node)) {
    throw new ManifestError("invalid_layout", "layout node must be a panel id or a row/col object");
  }
  const isRow = "row" in node;
  if (isRow === "col" in node) {
    throw new ManifestError("invalid_layout", "layout node needs exactly one of row/col");
  }
  const children = isRow ? node.row : node.col;
  if (!Array.isArray(children) || children.length === 0) {
    throw new ManifestError("invalid_layout", "row/col children must be a non-empty array");
  }
  let shares: number[] | undefined;
  if (node.shares !== undefined) {
    const raw = node.shares;
    // typeof excludes booleans; isFinite bounds to float64 like maxHz.
    if (
      !Array.isArray(raw) ||
      raw.length !== children.length ||
      !raw.every((s) => typeof s === "number" && Number.isFinite(s) && s > 0)
    ) {
      throw new ManifestError(
        "invalid_layout",
        "shares must be positive numbers, one per child",
      );
    }
    shares = [...(raw as number[])];
  }
  const rebuilt = children.map((child) => validateLayout(child, panelIds, seen));
  if (isRow) return shares === undefined ? { row: rebuilt } : { row: rebuilt, shares };
  return shares === undefined ? { col: rebuilt } : { col: rebuilt, shares };
}

/**
 * Validated manifest from parsed JSON (or any untrusted value); throws
 * ManifestError. Absent dir/params/title/layout/pages normalize to
 * "rx"/{}/""/null/[]; unknown keys are dropped (forward compatibility with
 * newer bridges). Mirrors parse_manifest() in manifest.py: shape, then
 * domain rules in document order, so both sides report the same code
 * (pinned by fixtures). The version gate runs before the channel/panel
 * shape checks so a structurally-alien future manifest still reports
 * unsupported_version, which the Cockpit turns into a polite notice.
 */
export function parseManifest(value: unknown): Manifest {
  if (!isRecord(value)) throw new ManifestError("invalid_shape", "manifest must be an object");
  if (typeof value.version !== "number") {
    throw new ManifestError("invalid_shape", "manifest version must be a number");
  }
  if (value.version !== MANIFEST_VERSION) {
    throw new ManifestError(
      "unsupported_version",
      `manifest version ${value.version} is not supported`,
    );
  }
  const rawChannels = value.channels;
  const rawPanels = value.panels === undefined ? [] : value.panels;
  if (
    !Array.isArray(rawChannels) || !rawChannels.every(isChannelSpec) ||
    !Array.isArray(rawPanels) || !rawPanels.every(isPanelSpec)
  ) {
    throw new ManifestError("invalid_shape", "bad channels/panels shape");
  }
  const channels = rawChannels as RawChannelSpec[];
  const panels = rawPanels as RawPanelSpec[];

  const chIds = new Map<string, RawChannelSpec>();
  for (const spec of channels) {
    if (!boundedId(spec.ch)) {
      throw new ManifestError(
        "invalid_channel_id",
        `channel id must be 1..${MAX_MANIFEST_ID_LEN} chars`,
      );
    }
    if (spec.ch.startsWith(RESERVED_CHANNEL_PREFIX)) {
      throw new ManifestError(
        "reserved_channel_id",
        `channel ids beginning with ${RESERVED_CHANNEL_PREFIX} are reserved for protocol control`,
      );
    }
    if (chIds.has(spec.ch)) {
      throw new ManifestError("duplicate_channel_id", `duplicate channel ${spec.ch}`);
    }
    chIds.set(spec.ch, spec);
    if (!boundedId(spec.encoding)) {
      throw new ManifestError(
        "invalid_encoding",
        `encoding must be 1..${MAX_MANIFEST_ID_LEN} chars`,
      );
    }
    // isFinite also bounds maxHz to float64: JSON.parse turns an
    // out-of-range literal into Infinity. manifest.py mirrors this with an
    // explicit float64 bound (Python parses such integers exactly).
    if (!Number.isFinite(spec.maxHz) || spec.maxHz <= 0) {
      throw new ManifestError("invalid_max_hz", `maxHz for ${spec.ch} must be a positive number`);
    }
    // Generic-publish rules. The manifest layer accepts "exclusive" (only
    // authoring and the bridge reject it until the lease ticket, W8).
    const publish = publishOf(spec);
    const scope = scopeOf(spec);
    if (dirOf(spec) === "rx" && publish !== "none") {
      throw new ManifestError("invalid_publish", `rx channel ${spec.ch} cannot declare publish`);
    }
    if (publish !== "none" && spec.delivery !== "reliable") {
      throw new ManifestError("invalid_publish", `publish channel ${spec.ch} must be reliable`);
    }
    if (publish !== "none" && !JSON_ENCODING_RE.test(spec.encoding)) {
      throw new ManifestError(
        "invalid_publish",
        `publish channel ${spec.ch} needs a JSON encoding`,
      );
    }
    if (scope !== null && publish === "none") {
      throw new ManifestError(
        "invalid_scope",
        `channel ${spec.ch} scope requires a publish policy`,
      );
    }
    if (scope !== null && !boundedId(scope)) {
      throw new ManifestError(
        "invalid_scope",
        `scope for ${spec.ch} must be 1..${MAX_MANIFEST_ID_LEN} chars`,
      );
    }
  }

  const panelIds = new Set<string>();
  for (const panel of panels) {
    if (!boundedId(panel.id) || !boundedId(panel.kind)) {
      throw new ManifestError(
        "invalid_panel",
        `panel id/kind must be 1..${MAX_MANIFEST_ID_LEN} chars`,
      );
    }
    if (panelIds.has(panel.id)) {
      throw new ManifestError("duplicate_panel_id", `duplicate panel ${panel.id}`);
    }
    panelIds.add(panel.id);
    for (const ch of panel.channels) {
      if (!chIds.has(ch)) {
        throw new ManifestError(
          "unknown_panel_channel",
          `panel ${panel.id} wants undeclared channel ${ch}`,
        );
      }
    }
    // Kind-specific rules (incl. dir=rx for viewer-rendered channels);
    // unknown kinds stay unvalidated (forward compatibility with newer
    // bridges).
    if (panel.kind === "video") {
      if (panel.channels.length !== 1) {
        throw new ManifestError(
          "invalid_video_panel",
          `video panel ${panel.id} must bind exactly one channel`,
        );
      }
      const bound = chIds.get(panel.channels[0])!;
      if (bound.encoding !== "jpeg.v1" || bound.delivery !== "latest" || dirOf(bound) !== "rx") {
        throw new ManifestError(
          "invalid_video_panel",
          `video panel ${panel.id} needs a jpeg.v1 latest rx channel`,
        );
      }
    }
    if (panel.kind === "map2d") {
      // channels[0] is the costmap; channels[1] (optional) the pose overlay.
      if (panel.channels.length !== 1 && panel.channels.length !== 2) {
        throw new ManifestError(
          "invalid_map2d_panel",
          `map2d panel ${panel.id} must bind one or two channels`,
        );
      }
      const costmap = chIds.get(panel.channels[0])!;
      if (
        costmap.encoding !== "costmap.zlib.v1" || costmap.delivery !== "latest" ||
        dirOf(costmap) !== "rx"
      ) {
        throw new ManifestError(
          "invalid_map2d_panel",
          `map2d panel ${panel.id} needs a costmap.zlib.v1 latest rx channel first`,
        );
      }
      if (panel.channels.length === 2) {
        const pose = chIds.get(panel.channels[1])!;
        if (pose.encoding !== "pose.json.v1" || dirOf(pose) !== "rx") {
          throw new ManifestError(
            "invalid_map2d_panel",
            `map2d panel ${panel.id} pose channel must be a pose.json.v1 rx channel`,
          );
        }
      }
    }
    if (panel.kind === "teleop") {
      if (panel.channels.length !== 1) {
        throw new ManifestError(
          "invalid_teleop_panel",
          `teleop panel ${panel.id} must bind exactly one channel`,
        );
      }
      const cmd = chIds.get(panel.channels[0])!;
      if (cmd.encoding !== "twist.json.v1" || cmd.delivery !== "latest" || dirOf(cmd) !== "tx") {
        throw new ManifestError(
          "invalid_teleop_panel",
          `teleop panel ${panel.id} needs a twist.json.v1 latest tx channel`,
        );
      }
    }
    if (panel.kind === "chat") {
      // channels: the text input (publish tx), the messages, the idle flag,
      // the push-to-talk audio (publish tx).
      if (panel.channels.length !== 4) {
        throw new ManifestError(
          "invalid_chat_panel",
          `chat panel ${panel.id} must bind four channels`,
        );
      }
      const [text, messages, idle, audio] = panel.channels.map((ch) => chIds.get(ch)!);
      if (
        text.encoding !== "text.json.v1" || text.delivery !== "reliable" ||
        dirOf(text) !== "tx" || publishOf(text) !== "shared"
      ) {
        throw new ManifestError(
          "invalid_chat_panel",
          `chat panel ${panel.id} needs a text.json.v1 reliable shared tx channel first`,
        );
      }
      if (
        messages.encoding !== "chat.json.v1" || messages.delivery !== "reliable" ||
        dirOf(messages) !== "rx"
      ) {
        throw new ManifestError(
          "invalid_chat_panel",
          `chat panel ${panel.id} needs a chat.json.v1 reliable rx channel second`,
        );
      }
      if (idle.encoding !== "json.v1" || idle.delivery !== "latest" || dirOf(idle) !== "rx") {
        throw new ManifestError(
          "invalid_chat_panel",
          `chat panel ${panel.id} needs a json.v1 latest rx channel third`,
        );
      }
      if (
        audio.encoding !== "audio.json.v1" || audio.delivery !== "reliable" ||
        dirOf(audio) !== "tx" || publishOf(audio) !== "shared"
      ) {
        throw new ManifestError(
          "invalid_chat_panel",
          `chat panel ${panel.id} needs an audio.json.v1 reliable shared tx channel fourth`,
        );
      }
    }
  }

  const seen = new Set<string>();
  let layout: LayoutNode | null = null;
  if (value.layout !== undefined && value.layout !== null) {
    layout = validateLayout(value.layout, panelIds, seen);
  }

  const rawPages = value.pages === undefined ? [] : value.pages;
  if (!Array.isArray(rawPages) || !rawPages.every((id) => typeof id === "string")) {
    throw new ManifestError("invalid_shape", "bad pages shape");
  }
  const pages = rawPages as string[];
  for (const id of pages) {
    if (!panelIds.has(id)) {
      throw new ManifestError("unknown_page_panel", `pages names unknown panel ${id}`);
    }
    if (seen.has(id)) {
      throw new ManifestError("duplicate_layout_panel", `panel ${id} placed more than once`);
    }
    seen.add(id);
  }

  return {
    version: MANIFEST_VERSION,
    channels: channels.map((spec) => ({
      ch: spec.ch,
      dir: dirOf(spec),
      encoding: spec.encoding,
      delivery: spec.delivery,
      maxHz: spec.maxHz,
      params: { ...(spec.params ?? {}) },
      publish: publishOf(spec),
      requiredScope: scopeOf(spec),
    })),
    panels: panels.map((panel) => ({
      id: panel.id,
      kind: panel.kind,
      title: panel.title ?? "",
      channels: [...panel.channels],
      params: { ...(panel.params ?? {}) },
    })),
    layout,
    pages: [...pages],
  };
}
